src/main/python/ttconv/config.py

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python
+# -*- coding: UTF-8 -*-
+
+# Copyright (c) 2020, Sandflow Consulting LLC
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""TT configuration"""
+
+from __future__ import annotations
+
+import dataclasses
+from dataclasses import dataclass
+from fractions import Fraction
+from typing import Optional, Dict, List, Any
+
+
+class ModuleConfiguration:
+  """Base class for module configurations"""
+
+  @classmethod
+  def get_fields(cls) -> List[dataclasses.Field]:
+    """Returns data class fields"""
+    return list(dataclasses.fields(cls))
+
+  @classmethod
+  def validate(cls, config_dict: Dict):
+    """Validates configuration dictionary"""
+    for field in cls.get_fields():
+      optional_field = "Optional" in field.type or cls.get_field_default(field) is not None
+
+      config_value = config_dict.get(field.name)
```

Comment:
What happens if half the fields are JSON types and half are custom types, e.g. fraction?

Reply:
Perhaps use `@dataclasses.field(metadata)` to store serializer/deserializer?

Reply:
See ff9011b